### PR TITLE
fix widget ids panels

### PR DESCRIPTION
**Pull request by Skalii**

### CODE REVIEW

File: Sources/AppHost/OsmAndAppImpl.mm [hunk 1]
```diff
@@ -301,6 +302,25 @@ - (void) migrateResourcesToDocumentsIfNeeded
         _resourcesManager->rescanUnmanagedStoragePaths(true);
 }
 
+- (void)migrateOldWidgetKeysIfNeeded
```
Comment (tigrim) [inline Sources/AppHost/OsmAndAppImpl.mm:305]:
```
@objcMembers final class StorageMigrationManager: NSObject {
    enum MigrationKey: String {
        case changeWidgetIdsMigration1Key
    }
    
    static let shared = StorageMigrationManager()
    let defaults = UserDefaults.standard
    
    private override init() {}
    
    func changeWidgetIdsMigration1() {
        guard defaults.object(forKey: MigrationKey.changeWidgetIdsMigration1Key.rawValue) == nil else { return }
        let changeWidgetIds = [
            "heartRate" : "ant_heart_rate",
            "bicycleCadence" : "ant_bicycle_cadence",
            "bicycleDistance" : "ant_bicycle_distance",
            "bicycleSpeed" : "ant_bicycle_speed",
            "temperature" : "temperature_sensor"
        ]
        // all logic of migration
        // ....
        defaults.set(true, forKey:  MigrationKey.changeWidgetIdsMigration1Key.rawValue)
    }
}
```

It's better to create a separate migration manager. The template is provided above

File: Sources/AppHost/OsmAndAppImpl.mm
Comment (tigrim) [inline Sources/AppHost/OsmAndAppImpl.mm:1]: - (OACommonInteger *)registerSizeStylePrefWith:(NSString *)customId
- (OACommonBoolean *)registerHideIconPrefWith:(NSString *)customId
need sync with android

File: Sources/Controllers/Map/Widgets/OATextInfoWidget.mm [hunk 2]
```diff
@@ -837,40 +837,34 @@ - (void)configurePrefsWithId:(NSString *)id appMode:(OAApplicationMode *)appMode
...
-    OAAppSettings *settings = [OAAppSettings sharedManager];
-    NSString *prefId = self.widgetType.title;
-    if (customId.length > 0)
-        prefId = [prefId stringByAppendingFormat:@"%@_%@",kSizeStylePref,customId];
-    else
-        prefId = [prefId stringByAppendingFormat:@"%@",kSizeStylePref];
-    
-    return [settings registerIntPreference:prefId defValue:WidgetSizeStyleMedium];
+    NSString *prefId = kSizeStylePref;
+    prefId = [prefId stringByAppendingString:self.widgetType.id];
```
Comment (tigrim) [inline Sources/Controllers/Map/Widgets/OATextInfoWidget.mm:855]: NSString *prefId  =  [kSizeStylePref stringByAppendingString:self.widgetType.id];

File: Sources/Controllers/Map/Widgets/OATextInfoWidget.mm [hunk 3]
```diff
@@ -837,40 +837,34 @@ - (void)configurePrefsWithId:(NSString *)id appMode:(OAApplicationMode *)appMode
...
-    OAAppSettings *settings = [OAAppSettings sharedManager];
-    NSString *prefId = self.widgetType.title;
-    if (customId.length > 0)
-        prefId = [prefId stringByAppendingFormat:@"%@_%@",kHideIconPref,customId];
-    else
-        prefId = [prefId stringByAppendingFormat:@"%@",kHideIconPref];
-    
-    return [settings registerBooleanPreference:prefId defValue:YES];
+    NSString *prefId = kShowIconPref;
+    prefId = [prefId stringByAppendingString:self.widgetType.id];
```
Comment (tigrim) [inline Sources/Controllers/Map/Widgets/OATextInfoWidget.mm:864]: up

File: Sources/Helpers/MigrationManager.swift [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+final class MigrationManager: NSObject {
+
+    enum MigrationKey: String {
+        case migrationChangeWidgetIds1Key
+    }
+
+    static let shared = MigrationManager()
+    let defaults = UserDefaults.standard
+
+    let changeWidgetIds1 = [
```
Comment (tigrim) [inline Sources/Helpers/MigrationManager.swift:22]: declare in changeWidgetIdsMigration func

File: Sources/Helpers/MigrationManager.swift [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        "heartRate": "ant_heart_rate",
+        "bicycleCadence": "ant_bicycle_cadence",
+        "bicycleDistance": "ant_bicycle_distance",
+        "bicycleSpeed": "ant_bicycle_speed",
+        "temperature": "temperature_sensor"
+    ]
+
+    private override init() {}
+
+    func changeWidgetIdsMigration1(_ firstLaunch: Bool) {
```
Comment (tigrim) [inline Sources/Helpers/MigrationManager.swift:32]: private func

Comment (tigrim) [reply Sources/Helpers/MigrationManager.swift:32]: remove firstLaunch:

File: Sources/Helpers/MigrationManager.swift [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        if firstLaunch {
+            defaults.set(true, forKey: MigrationKey.migrationChangeWidgetIds1Key.rawValue)
+            return
+        }
+
+        let isOldWidgetKeysMigrated = defaults.object(forKey: MigrationKey.migrationChangeWidgetIds1Key.rawValue)
```
Comment (tigrim) [inline Sources/Helpers/MigrationManager.swift:38]: guard UserDefaults.standard.bool(forKey: "MigrationKey.migrationChangeWidgetIds1Key.rawValue") else { return }

File: Sources/Helpers/MigrationManager.swift [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+                                        panelPreference: settings.rightWidgetPanelOrder)
+
+                updateCustomWidgetKeys(mode, changeWidgetIds: changeWidgetIds1)
+                updateMapInfoControls(mode, changeWidgetIds: changeWidgetIds1)
+            }
+            defaults.set(true, forKey: MigrationKey.migrationChangeWidgetIds1Key.rawValue)
+        }
+    }
+
+    func updateExistingWidgetIds(_ appMode: OAApplicationMode,
```
Comment (tigrim) [inline Sources/Helpers/MigrationManager.swift:68]: private

File: Sources/Helpers/MigrationManager.swift [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        }
+        if pages != newPages {
+            panelPreference.set(newPages, mode: appMode)
+        }
+        if let newPanelPreference {
+            newPanelPreference.set(newPages, mode: appMode)
+        }
+    }
+
+    func updateCustomWidgetKeys(_ appMode: OAApplicationMode, changeWidgetIds: [String: String]) {
```
Comment (tigrim) [inline Sources/Helpers/MigrationManager.swift:89]: private

File: Sources/Helpers/MigrationManager.swift [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        guard let customIds = customWidgetKeys.get(appMode),
+              (customIds.contains { changeWidgetIds.keys.contains(WidgetType.getDefaultWidgetId($0)) }) else { return }
+
+        let newCustomIds = getUpdatedWidgetIds(customIds, changeWidgetIds: changeWidgetIds)
+        if customIds != newCustomIds {
+            customWidgetKeys.set(newCustomIds, mode: appMode)
+        }
+    }
+
+    func updateMapInfoControls(_ appMode: OAApplicationMode, changeWidgetIds: [String: String]) {
```
Comment (tigrim) [inline Sources/Helpers/MigrationManager.swift:100]: private

File: Sources/Helpers/MigrationManager.swift [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        let widgetsVisibility = widgetsVisibilityString.components(separatedBy: SETTINGS_SEPARATOR)
+        guard (widgetsVisibility.contains { changeWidgetIds.keys.contains(WidgetType.getDefaultWidgetId($0)) }) else { return }
+
+        let newWidgetsVisibility = getUpdatedWidgetIds(widgetsVisibility, changeWidgetIds: changeWidgetIds)
+        if widgetsVisibility != newWidgetsVisibility {
+            mapInfoControls.set(newWidgetsVisibility.joined(separator: SETTINGS_SEPARATOR), mode: appMode)
+        }
+    }
+
+    func getUpdatedWidgetIds(_ widgetIds: [String], changeWidgetIds: [String: String]) -> [String] {
```
Comment (tigrim) [inline Sources/Helpers/MigrationManager.swift:113]: private

File: Sources/Helpers/MigrationManager.swift (hunk 4 above)
Comment (tigrim) [inline Sources/Helpers/MigrationManager.swift:22]: changeWidgetIds

File: Sources/Helpers/OAAppSettings.h [hunk 11]
```diff
@@ -970,10 +970,13 @@ typedef NS_ENUM(NSInteger, EOARateUsState)
 // Widgets
 
 @property (nonatomic) OACommonListOfStringList *leftWidgetPanelOrder;
-@property (nonatomic) OACommonListOfStringList *topWidgetPanelOrder;
 @property (nonatomic) OACommonListOfStringList *rightWidgetPanelOrder;
+@property (nonatomic) OACommonListOfStringList *topWidgetPanelOrder;
 @property (nonatomic) OACommonListOfStringList *bottomWidgetPanelOrder;
 
+@property (nonatomic) OACommonListOfStringList *topWidgetPanelOrderOld;
+@property (nonatomic) OACommonListOfStringList *bottomWidgetPanelOrderOld;
```
Comment (tigrim) [inline Sources/Helpers/OAAppSettings.h:978]: Can we refrain from using the new properties topWidgetPanelOrderOld and bottomWidgetPanelOrderOld and instead override the old topWidgetPanelOrder/bottomWidgetPanelOrder in the changeWidgetIdsMigration1 function?

File: Sources/Helpers/MigrationManager.swift (hunk 5 above)
Comment (tigrim) [reply Sources/Helpers/MigrationManager.swift:32]: migrateIfNeeded

File: Sources/Helpers/MigrationManager.swift [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+            customWidgetKeys.set(newCustomIds, mode: appMode)
+        }
+    }
+
+    private func updateMapInfoControls(_ appMode: OAApplicationMode, changeWidgetIds: [String: String]) {
+        let mapInfoControls: OACommonString = OAAppSettings.sharedManager().mapInfoControls
+        guard let widgetsVisibilityString: String = mapInfoControls.get(appMode) else { return }
+
+        let widgetsVisibility = widgetsVisibilityString.components(separatedBy: SETTINGS_SEPARATOR)
+        guard (widgetsVisibility.contains { changeWidgetIds.keys.contains(WidgetType.getDefaultWidgetId($0)) }) else { return }
```
Comment (tigrim) [inline Sources/Helpers/MigrationManager.swift:123]: remove () ->  guard widgetsVisibility.contains

File: Sources/Helpers/Extensions/OAAppSettings+Extension.swift [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+//  OsmAnd Maps
+//
+//  Created by Skalii on 15.03.2024.
+//  Copyright © 2024 OsmAnd. All rights reserved.
+//
+
+import Foundation
+
+extension OACommonInteger {
+    @objc func stringToEnum(value: String) -> Int32 {
```
Comment (tigrim) [inline Sources/Helpers/Extensions/OAAppSettings+Extension.swift:12]: stringToEnum but return Int32

File: Sources/Helpers/Extensions/OAAppSettings+Extension.swift [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+        if key.hasPrefix("simple_widget_size") {
+            return Int32(WidgetSizeStyle.getValueFromString(value))
+        }
+        return Int32(value) ?? 0
+    }
+
+    @objc func enumToString(mode: OAApplicationMode) -> String {
```
Comment (tigrim) [inline Sources/Helpers/Extensions/OAAppSettings+Extension.swift:19]: getStringForAppMode

File: Sources/Helpers/MigrationManager.swift [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+    enum MigrationKey: String {
+        case migrationChangeWidgetIds1Key
+    }
+
+    static let shared = MigrationManager()
+    let defaults = UserDefaults.standard
+
+    private override init() {}
+
+    func migrateIfNeeded(_ firstLaunch: Bool) {
```
Comment (tigrim) [inline Sources/Helpers/MigrationManager.swift:24]: isFirstLaunch

File: Sources/Helpers/MigrationManager.swift [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+                BLE sensor widget IDs:
+
+                heartRate -> ant_heart_rate
+                bicycleCadence -> ant_bicycle_cadence
+                bicycleDistance -> ant_bicycle_distance
+                bicycleSpeed -> ant_bicycle_speed
+                temperature -> temperature_sensor */
+
+            let isOldWidgetKeysMigrated = defaults.object(forKey: MigrationKey.migrationChangeWidgetIds1Key.rawValue)
+            if isOldWidgetKeysMigrated == nil || (isOldWidgetKeysMigrated as! Bool) == false {
```
Comment (tigrim) [inline Sources/Helpers/MigrationManager.swift:45]: use open func bool(forKey defaultName: String) -> Bool (defaults)

File: Sources/Helpers/MigrationManager.swift [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+
+                        let useAnyDevicePrefKey = "\(widgetType.title)_useAnyDevicePref_\(prefKeySuffix)_\(appModeStringKey)"
+                        if let useAnyDevicePref = defaults.object(forKey: useAnyDevicePrefKey) as? Bool,
+                           let plugin = OAPlugin.getByWidgetId(newOriginalWidgetId),
+                           let fieldType = plugin.getWidgetDataFieldTypeName(byWidgetId: newOriginalWidgetId) {
+                            let newPrefKey = fieldType + (hasCustomSuffix ? newCustomWidgetId : "")
+                            let newPref: OACommonString = settings.registerStringPreference(newPrefKey, defValue: plugin.getAnyConnectedDeviceId())
+                            if !useAnyDevicePref {
+                                let oldPrefKey = "\(widgetType.title)\(oldCustomWidgetId)_\(appModeStringKey)"
+                                if let oldPref = defaults.object(forKey: oldPrefKey) as? String {
```
Comment (tigrim) [inline Sources/Helpers/MigrationManager.swift:153]: open func string(forKey defaultName: String) -> String?

File: Sources/Plugins/ExternalSensorsPlugin/BLE/BLEManager/SensorTextWidget.swift [hunk 18]
```diff
@@ -15,24 +15,23 @@ final class SensorTextWidget: OASimpleWidget {
...
-    
     private var cachedValue: String?
-    private var deviceIdPref: OACommonPreference?
+    private var deviceIdPref: OACommonString?
     private var appMode: OAApplicationMode!
+    private var plugin: OAExternalSensorsPlugin?
     
     var shouldUseAnyDevice: Bool {
-        useAnyDevicePref?.get() == true
+        deviceIdPref?.get(appMode) == plugin?.getAnyConnectedDeviceId()
```
Comment (tigrim) [inline Sources/Plugins/ExternalSensorsPlugin/BLE/BLEManager/SensorTextWidget.swift:24]: In which case will it be false?

**Review by tigrim — APPROVED** [review]

File: Sources/Helpers/Extensions/OAAppSettings+Extension.swift [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+//
+//  Created by Skalii on 15.03.2024.
+//  Copyright © 2024 OsmAnd. All rights reserved.
+//
+
+import Foundation
+
+extension OACommonInteger {
+    @objc func getValueFromString(_ value: String) -> Int32 {
+        if key.hasPrefix("simple_widget_size") {
```
Comment (alex-osm) [inline Sources/Helpers/Extensions/OAAppSettings+Extension.swift:13]: Value of `OACommonInteger` should not have any string inside

File: Sources/Helpers/OAAppSettings.m [hunk 20]
```diff
@@ -1781,12 +1784,12 @@ - (void) resetToDefault
 
 - (void)setValueFromString:(NSString *)strValue appMode:(OAApplicationMode *)mode
 {
-    [self set:strValue.intValue mode:mode];
+    [self set:[self getValueFromString:strValue] mode:mode];
```
Comment (alex-osm) [inline Sources/Helpers/OAAppSettings.m:1787]: Revert

File: Sources/Helpers/OAAppSettings.m [hunk 21]
```diff
@@ -1781,12 +1784,12 @@ - (void) resetToDefault
...
 }
 
 - (NSString *)toStringValue:(OAApplicationMode *)mode
 {
-    return [NSString stringWithFormat:@"%d", [self get:mode]];
+    return [self getStringFromValue:mode];
```
Comment (alex-osm) [inline Sources/Helpers/OAAppSettings.m:1792]: Revert

**Pull request merged**